### ISSUE TITLE
Namespace issues

### ISSUE THREAD

**Issue by jaytaph** [saffire/saffire#270]
I don't seem to get the namespaces right, no matter how hard I try.. I seem to miss obvious things, but I'm already tunneled, so I really need some other insights in the matter..

Right now, namespaces are implicit. You don't add a namespace (a la PHP) to your code, but the directory where you reside is basically the namespace (provided some rules).

The main application will always have the root namespace, as defined by `\`  (a backslash, it WILL become forward to imply they are linked to "filepaths", but currently has some lexing/parsing issues which I solve later. You can see \ and / being similar for now).

so running:

```
 `/home/jthijssen/test.sf`
```

results in the whole file `test.sf` to have the namespace `\`.
## Importing modules

A module is nothing more than a directory. This directory contains a saffire file (normally a simple `.sf` file, but it could also be just the bytecode `.sfc`, or even a C-compiled module in the form of a shared object (`.so`) file.

Importing works in the following way:

```
 import <qcn> as <alias> from <module>
```

where:
- **qcn**  A qualified class name in the form of either `foo` or `foo\bar`. Thus the qualified namespace can have a namespaced name in it.
- **alias** An identifier that CANNOT be namespaced. This is how you would reference the imported class within the current module. You could import the same class with different aliases, and both will reference to the same class.
- **module** A qualified class name. This will be the module where the class can be found.

Examples:

```
import response as response from http;
```

This will import the `response` **class**, from the http **module**. Inside the current class (actually, the current context), it will be known as the **alias** `response`.

If the **qcn** and the **alias** are the same, you can omit the **alias**. These two imports are equal:

```
import response as response from http;
import response from http;
```

Also, when you use a namespaced **qcn**, the BASECLASS is used as the **alias** when omitted. Both these imports are equal:

```
 import http/response as response from framework;
 import http/response from framework;
```

**_Question: Should we allow QCN's or just single names just like in the aliases? What do we gain/loose when making this strict?**_

Not only can we omit the **alias**, we could also omit the **module** if the **module** is the same name as the **qcn**. The following three import statements are equal:

```
  import response as response from response;
  import response from response;
  import response;
```

Also, when using a QCN, the following pairs are equal:

```
 import http/response;
 import response as response from http;

 import framework/http/response;
 import response as response from framework/http;

 import framework/http/response as foo;
 import response as foo from framework/http;
```

It's quite common to use the shorthand notation unless you want to change the **alias** of the given class, in case of conflicts with other already imported aliases. In that case, you could either use the full form, or just the **class** and **alias** form.

**\* Question: should we just skill the module part, and only allow qcn and alias. Since, we can easily "merge" the module and qcn together as we do in the short notation.***

**\* Question: Because of these shortcut notation, we must allow namespaced QCNs. If we don't allow QCNs when supplying a from, we surely must allow namespaces QCNs when omitting the from part.. Should this be considered an inconsistency?***
## Resolving classes and namespaces

Resolving namespaces only takes place on the first call to the given alias. Resolving a class works in a few different ways. Once the **alias** is referenced the following takes place:
- parse the baseclass and the module path from the given namespace. For instance, if the alias `response` resolves to `\framework\http\response`, the **class** name will become `response`, and the module-path `\framework\http`.
- convert the module path to an FQCN. An FQCN is a name that starts with `\` and are not context-dependend. If a class is not FQCN, the system will prepend the current namespace context to the context. Thus, a namespace `framework\http` (with the leading `\`), is NOT a FQCN, and thus will get the current context prepended. In the main saffire file, it will be `\` and becomes `\framework\http`. In a file who's context happens to be `\foo\bar` (context are always FQCN), the module path would be come `\foo\bar\framework\http`.
- Convert module path to file path. On the final system, this should not be needed, as module paths and file paths follow the same system. However, for now, it means that all `\`, must be changed to `/`, so `\framework\http`, becomes `/framework/http`.
- Resolve module-path + class through module search paths. (see below).
- Once a class has been found and imported without and exceptions, it will check for existence of the needed class (`response` in this case). If not, it will throw an import exception, otherwise this response class is linked to the alias of the calling context.
### Resolving paths

Once the module path is found and convert to a file path, there are different ways to find and search the actual file.

First of all, the destination file of what to look for is:   `<file_path>/<response>`. Thus the target file to find will be `/framework/http/response`.

Since there can be different types and extensions that can be used, it will look in the following way:
- Search for `/framework/http/response.sf`  (saffire source)
- Search for `/framework/http/response.sfc` (saffire bytecode)
- Search for `/framework/http/response.so` (saffire binary module)

There are also different source paths where modules itself can reside. These can be changed by configuration options, but for now are hardcoded:
- Search in the current directory (`.`) of the MAIN CALLED FILE,
- Search in the project modules directory (`./saffire_modules`) of the MAIN CALLED FILE
- Search in the global modules directory (`/usr/lib/saffire/modules`).

So, resolving the class will check the following, provided the main saffire class is `/home/jthijssen/test.sf`:

```
   #1 /home/jthijssen/./framework/http/response.sf
   #2 /home/jthijssen/./framework/http/response.sfc
   #3 /home/jthijssen/./framework/http/response.so
   #4 /home/jthijssen/./saffire_module/framework/http/response.sf
   #5 /home/jthijssen/./saffire_module/framework/http/response.sfc
   #6 /home/jthijssen/./saffire_module/framework/http/response.so
   #7 /usr/lib/saffire/modules/framework/http/response.sf
   #8 /usr/lib/saffire/modules/framework/http/response.sfc
   #9 /usr/lib/saffire/modules/framework/http/response.so
```
### Resolving internal modules

Saffire provides some own modules, all residing in the reservered `\saffire` namespace. For instance `\saffire\io`, `\saffire\saffire`, `\saffire\os`, `\saffire\io\socket`.

There is always one single class form the `\saffire\saffire` module loaded: `\saffire`, which is also aliased to `saffire`. Thus, it's always possible to call `saffire.version()` (and other methods from the `saffire` class). This is needed, because an exception will call the `saffire.uncaughtExceptionHandler()` method and must always be present (otherwise you will get an exception during handling an exception).

Note that since (most) classes are internal, they do not need to be resolved, as they are already known on their FQCN by the system.

Let's assume that all our internal modules are namespaced within `\saffire`. The `IO` module consists of two classes: `io` and `console`. One being generic IO, the other for plain console IO (not really used for now). Thus using the `IO` **class** from the `IO` **module** would become:  

```
import io as io from \saffire\io;
```

or, shorthand:

```
import \saffire\io;
```

However, due to historic reasons, the following will work too:

```
import io;
import io as io from io;
```

What happens here, is that when a module cannot be resolved through any of the above (9) paths, it will retry again, but this time with the `\saffire` prepended in front of the module path. Thus, after not being able to find the  `/io/io.sf`, `/io/io.sfc` etc, it will look for a module / class path:

```
`\saffire\io\io`.
```

Before doing the actual file resolve, it will check if this class is already known inside Saffire. When the modules are internal (which they are, in case of the `IO` module), it means that it does not have to resolve the paths, but simply can use the internal module. Otherwise, it will redo the 9 file path checks, but this time with `/saffire` prepended.

Note: Current it will also check for `_sfl` suffix. This is not used anymore and must be removed.

**\* Question: should we really allow this kind of "implicit" searches? It makes the import statements smaller if we do, but we run into name collisions easier: IO for instance, is something that can easily be used for third party modules. Doing an "import io;" would then import the module's IO class, not the Saffire one.
## Multi import lines

When importing, the following can happen:

```
  import io;
  import os;
  import http/request;
  import http/response;
```

Sometimes, a lot of small imports take place, all eating up a single line. The following notation ispossible too:

```
 import io, os, http/request, http/response;
```

A comma separated list of different imports. However, there are some complications with this:

```
import io as foo, os as bar from baz;
```

Does this imply `import io as foo from io`, and `os as bar from baz` OR  `import io as foo from baz` and `importos as bar from baz`.

Basically, which takes precedence with comma's:

```
import (io as foo, os as bar) from baz;
```

or

```
import (io as foo), (os as bar from baz);
```

**\* Question: if we do not have a "from" part, because we only use the QCN, does it not solve this issue too? ***

**Comment by jaytaph** [comment]
When removing the `from` part, it feels a bit better, and takes away the parsing ambiguity:

```
import io,os;
// import io as io from io;   import os as os from os;

import io as foo, os as bar;
// import io as foo from io;   import os as bar from os;

import http/request as req, http/response as res;
// import request as req from http/request;
// import response as res from http/response;

import http/request, http/response as res;
// import request as request from http/request;
// import response as res from http/response;
```
